velox/exec/Task.cpp

```diff
@@ -1546,14 +1546,21 @@ ContinueFuture Task::terminate(TaskState terminalState) {
   std::
       unordered_map<core::PlanNodeId, std::pair<std::vector<exec::Split>, bool>>
           remainingRemoteSplits;
+  std::vector<ContinuePromise> barrierPromises;
   {
     std::lock_guard<std::mutex> l(mutex_);
+
     // Collect all the join bridges to clear them.
     for (auto& splitGroupState : splitGroupStates_) {
       for (auto& pair : splitGroupState.second.bridges) {
         oldBridges.emplace_back(std::move(pair.second));
       }
       splitGroupStates.push_back(std::move(splitGroupState.second));
+
+      // Collect all outstanding barrier promises.
```

Comment:
The barrier is cleared with split group state at l1602? Is it required?

Reply:
They are cleared, but not fulfilled.
We have broken promise messages all over the logs and they point to the barriers.
This logic is to fulfill the outstanding promises.
I suspect this might be the reason for "std::exception" message instead of "memory limit reached" errors in some queries.

Reply:
I see. Shall we call this splitBarrierPromises?

Reply:
These are not split barriers.
These are barriers for certain logic that requires the last entity to do some work for everyone.
It lives in split group related structure just for separation between split groups.
The current name is fine, I believe.

Reply:
Shall we add to check if a task has been terminated in allPeersFinished() to avoid access the promises in the split group state? std::exception might be caused by that? Task terminate will clear the objects inside the split group state and a late running driver might access the split group state inside allPeersFinished()?

Reply:
@xiaoxmeng 
We already have such check
```
  if (exception_) {
    VELOX_FAIL(
        "Task is terminating because of error: {}",
        errorMessageImpl(exception_));
  }
```

I'm not sure where "std::exception" comes from and it is not reproducible.
I noticed broken promise messages around when we had such incident and want to fix that.
Then we'll see if the "std::exception" goes away or not.
In the latter case we'll continue investigation.
Accessing split group state after it was cleared up is ok - it will be recreated in the maps and will be completely empty.
But we should really exit before due to the check above.
I don't think allPeersFinished() causes any issues here and the change is not about it.
Hopefully this explains the situation.